// @startx/probot-memory/src/business-mapping/property.js
"use strict";
const Sequelize = require('sequelize');
var common = require("./common.json");
var PropertySchema = {
    identifier: {
        type: Sequelize.STRING,
        primaryKey: true,
        allowNull: false,
        unique: true,
    },
    words: {
        type: Sequelize.JSON,
        allowNull: false,
    },
    value: {
        type: Sequelize.JSON,
        allowNull: true,
    },
    entity: {
        type: Sequelize.STRING,
        allowNull: false,
    },
    type: {
        type: Sequelize.ENUM('MONEY', 'DATE', 'NUMBER', 'OPEN_STRING', 'PERCENT', 'EMAIL', 'PHONE', 'ADDRESS', 'NAME', 'CONTACT', 'POINTER', 'RADIOBOX'),
        allowNull: false,
    },
    visible: {
        type: Sequelize.BOOLEAN,
        allowNull: false,
        defaultValue: 0
    },
    order: {
        type: Sequelize.INTEGER,
        allowNull: true,
        defaultValue: 0
    },
    friendlyName: {
        type: Sequelize.TEXT('tiny'),
        allowNull: false,
    },
    businessType: {
        type: Sequelize.JSON,
        allowNull: false,
    },
    collectionName: {
        type: Sequelize.STRING,
        allowNull: false,
    },
};
var table = "properties";
module.exports = {
    table,
    schema: PropertySchema,
};
