// @startx/probot-memory/src/business-mapping/intent.js
"use strict";
const Sequelize = require('sequelize');
var common = require("./common.json");
var IntentSchema = {
    id: {
        type: Sequelize.INTEGER,
        allowNull: false,
        autoIncrement: true,
        primaryKey: true
    },
    category: {
        type: Sequelize.ENUM('GET', 'DELETE', 'UPDATE', 'CREATE', 'RULE'),
        allowNull: false,
    },
    subCategory: {
        type: Sequelize.ENUM('SOLE', 'PLURAL', 'YES_NO', 'VALUE', 'FORMULA'),
        allowNull: false,
    },
    words: {
        type: Sequelize.JSON,
        allowNull: false,
    },
    accountType: {
        type: Sequelize.ENUM('ASANA', 'GDRIVE', 'GMAIL', 'OTHER', 'HUBSPOT', 'SALESFORCE', 'ZENDESK'),
        allowNull: false,
    },
    applicationType: {
        type: Sequelize.ENUM('CRM', 'STORAGE', 'EMAIL', 'GLOBAL', 'HELPDESK'),
        allowNull: false,
    },
    businessType: {
        type: Sequelize.JSON,
        allowNull: false,
    },
};
var table = "intents";
module.exports = {
    table,
    schema: IntentSchema,
};
