// @startx/probot-memory/src/business-mapping/skill.js
"use strict";
const Sequelize = require('sequelize');
var common = require("./common.json");
var SkillSchema = {
    identifier: {
        type: Sequelize.STRING,
        primaryKey: true,
        allowNull: false,
        unique: true,
    },
    description: {
        type: Sequelize.TEXT,
        allowNull: false,
    },
    words: {
        type: Sequelize.JSON,
        allowNull: false,
    },
    endpoint: {
        type: Sequelize.TEXT,
        allowNull: true,
    },
    mainEntity: {
        type: Sequelize.STRING,
        allowNull: false,
    },
    mainIntent: {
        type: Sequelize.STRING,
        allowNull: false,
    },
    businessType: {
        type: Sequelize.JSON,
        allowNull: false,
    },
    accountType: {
        type: Sequelize.ENUM('ASANA', 'GDRIVE', 'GMAIL', 'OTHER', 'HUBSPOT', 'SALESFORCE', 'ZENDESK'),
        allowNull: false,
    },
    applicationType: {
        type: Sequelize.ENUM('CRM', 'STORAGE', 'EMAIL', 'GLOBAL', 'HELPDESK'),
        allowNull: false,
    },
};
var table = "skills";
module.exports = {
    table,
    schema: SkillSchema,
};
